packages/toolpad-app/src/toolpad/DomLoader.tsx

```diff
@@ -75,8 +82,45 @@ export type DomAction =
       node: appDom.AppDomNode;
     };
 
+const undoStack: appDom.AppDom[] = [];
```

Comment:
Perhaps we can
- add the undostack to the `DomLoader` state and keep it immutable? We're basically updating mutable global state in a reducer here. Which is sort of the opposite of what a reducer should do.
- limit the length of the stack to avoid growing indefinitely?

Reply:
@Janpot so I've tried to move data to the state itself (attempt can be seen here: https://github.com/mui/mui-toolpad/pull/1225/commits/ff09a8220658878e9237e515bf6f704153fd57f5 ) at first I didn't understand why it lead to unpredictable behaviour, however later I discovered the problem with storing array of refs in state and the biggest problem was that I had to mutate that array from within throttled function which basically beats the purpose of having it in state since I can't derive new state properly.

If we already had atomic updates, then I would change the implementation so we don't use array of "check points" at all and instead IMO we should store the pointer to the previous and next state which would probably be the most straight forward implementation 🤔 

As for second point - I added a limit of undo actions

Reply:
> so I've tried to move data to the state itself 

[This](https://github.com/mui/mui-toolpad/blob/ff09a8220658878e9237e515bf6f704153fd57f5/packages/toolpad-app/src/toolpad/DomLoader.tsx#L159) is still doing side-effects in a reducer. 


I would try the following:

- Add an action `UNDO_HISTORY_PUSH` which reduces by taking the current dom and pushing it on the undo stack
- create a `scheduleUndoHistoryPush ` method that dispatches this action and wrap the `dispatch` method:

   ```tsx
    const scheduleUndoHistoryPush = useThrottled(() => {
      dispatch({ type: 'UNDO_HISTORY_PUSH' })
    }, [500])

    const dispatchWithUndo = React.useCallback((action) => {
      dispatch(action);
      if (isUndoable(action)) {
        scheduleUndoHistoryPush()
      }
    }, [dispatch, isUndoable, scheduleUndoHistoryPush])
   ```
- Use this `dispatchWithUndo` wrapper in the `domApi` instead of `dispatch`

Reply:
Thanks alot! That's great tip and I think I finally got it all working 😅 @Janpot please take a look when you can.

There is one thing that I couldn't figure out - I had to use 
```
// eslint-disable-next-line react-hooks/exhaustive-deps
const api = React.useMemo(() => createDomApi(dispatchWithHistory), []);
```
If I pass dependency as eslint requires everything seems to blow up and entire app works in unpredictable way 🤔

Reply:
Yes, looking good. What happens if you wrap `dispatchWithHistory` with `useEvent`? (We can't call it during render then, but I don't think we do that anywhere, and we probably never should anyway.)

Reply:
Are you talking about this `useEvent`: https://beta.reactjs.org/apis/react/useEvent ? I think it's not supported yet in the current version?

Reply:
It's a common practice to have such hook in your codebase: https://github.com/mui/mui-toolpad/blob/9c8764b689dbb7a44cc8d6833ffd393bdf402cd6/packages/toolpad-app/src/utils/useEvent.ts#L24

This version also contains an additional check that prevents us from accidentally using it during render.

Reply:
> It's a common practice to have such hook in your codebase

```
__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentDispatcher
```

Is it really a common practice? I mean you can't be more explicit than that to no use these internals 🤔

Reply:
https://github.com/mui/mui-toolpad/pull/1269

Reply:
@Janpot so instead of using `useMemo` I tried using `useEvent` like so:
```diff
diff --git a/packages/toolpad-app/src/toolpad/DomLoader.tsx b/packages/toolpad-app/src/toolpad/DomLoader.tsx
index ed031d9d..370920e8 100644
--- a/packages/toolpad-app/src/toolpad/DomLoader.tsx
+++ b/packages/toolpad-app/src/toolpad/DomLoader.tsx
@@ -11,6 +11,7 @@ import { createProvidedContext } from '../utils/react';
 import { mapValues } from '../utils/collections';
 import insecureHash from '../utils/insecureHash';
 import { NodeHashes } from '../types';
+import useEvent from '../utils/useEvent';
 
 export type DomAction =
   | {
@@ -414,16 +415,14 @@ export default function DomProvider({ appId, children }: DomContextProps) {
     redoStack: [],
   });
 
-  const scheduleHistoryUpdate = React.useMemo(
-    () =>
-      throttle(
-        () => {
-          dispatch({ type: 'DOM_UPDATE_HISTORY' });
-        },
-        500,
-        { leading: false, trailing: true },
-      ),
-    [],
+  const scheduleHistoryUpdate = useEvent(() =>
+    throttle(
+      () => {
+        dispatch({ type: 'DOM_UPDATE_HISTORY' });
+      },
+      500,
+      { leading: false, trailing: true },
+    ),
   );
 
   const dispatchWithHistory = (action: DomAction) => {

```

But I either do that incorrectly or it's not working for some reason as undo just does nothing after that 🤔

Reply:
```tsx
  const dispatchWithHistory = useEvent((action: DomAction) => {
    dispatch(action);

    if (!SKIP_UNDO_ACTIONS.has(action.type)) {
      scheduleHistoryUpdate();
    }
  });

  const api = React.useMemo(() => createDomApi(dispatchWithHistory), [dispatchWithHistory]);
```

Reply:
🤦 ah, my bad... thanks! now it's working 👍